lib/instrumentation/modules/@aws-sdk/client-sns.js

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright Elasticsearch B.V. and other contributors where applicable.
+ * Licensed under the BSD 2-Clause License; you may not use this file except in
+ * compliance with the BSD 2-Clause License.
+ */
+
+'use strict';
+
+const constants = require('../../../constants');
+const NAME = 'SNS';
+const TYPE = 'messaging';
+const SUBTYPE = 'sns';
+const MAX_SNS_MESSAGE_ATTRIBUTES = 10;
+const elasticAPMStash = Symbol('elasticAPMStash');
+
+/**
+ * Returns middlewares to instrument an S3Client instance
+ *
+ * @param {import('@aws-sdk/client-sns').SNSClient} client
+ * @param {any} agent
+ * @returns {import('./smithy-client').AWSMiddlewareEntry[]}
+ */
+function snsMiddlewareFactory(client, agent) {
+  return [
+    {
+      middleware: (next, context) => async (args) => {
+        const ins = agent._instrumentation;
+        const log = agent.logger;
+        const span = ins.currSpan();
+        const input = args.input;
+
+        // W3C trace-context propagation.
+        const runContext = ins.currRunContext();
+        const parentSpan =
+          span || runContext.currSpan() || runContext.currTransaction();
+        const targetId =
+          input && (input.TopicArn || input.TargetArn || input.PhoneNumber);
+
+        // Though our spec only mentions a 10-message-attribute limit for *SQS*, we'll
+        // do the same limit here, because
+        // https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
+        // mentions the 10-message-attribute limit for SQS subscriptions.
+        input.MessageAttributes = input.MessageAttributes || {};
+        if (
+          input.MessageAttributes &&
+          Object.keys(input.MessageAttributes).length + 2 >
+            MAX_SNS_MESSAGE_ATTRIBUTES
+        ) {
+          log.warn(
+            'cannot propagate trace-context with SNS message to %s, too many MessageAttributes',
+            targetId,
+          );
+        } else {
+          parentSpan.propagateTraceContextHeaders(
+            input.MessageAttributes,
+            function (msgAttrs, name, value) {
+              if (name.startsWith('elastic-')) {
+                return;
+              }
+              msgAttrs[name] = { DataType: 'String', StringValue: value };
+            },
+          );
+        }
+
+        // Ensure there is a span from the wrapped `client.send()`.
+        if (!span || !(span.type === TYPE && span.subtype === SUBTYPE)) {
+          return await next(args);
+        }
+
+        let queueName;
```

Comment:
nit: Perhaps `topicName` instead of `queueName`. However, this is really minor. `queueName` is just fine. I see that `aws-sdk/sns.js` used both names.  You are being consistent here, which is better.

Reply:
changed to `topicName`